astropy/modeling/tests/test_quantities.py

```diff
@@ -0,0 +1,232 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""Tests specifically for models that use units and quantities."""
+
+from __future__ import (absolute_import, unicode_literals, division,
+                        print_function)
+
+import numpy as np
+from numpy.testing import assert_allclose
+
+
+from ..core import Fittable1DModel, ModelDefinitionError, InputParameterError
+from ..parameters import Parameter, ParameterDefinitionError
+from ..models import Gaussian1D
+from ... import units as u
+from ...units import UnitsError, Quantity
+from ...tests.helper import pytest
+
+
+def test_quantities_as_parameters():
+    """
+    Basic tests for initializing general models (that do not require units)
+    with parameters that have units attached.
+    """
+
+    g = Gaussian1D(1 * u.J, 1 * u.m, 0.1 * u.m)
+    assert g.amplitude.value == 1.0
+    assert g.amplitude.unit is u.J
+    assert g.mean.value == 1.0
+    assert g.mean.unit is u.m
+    assert g.stddev.value == 0.1
+    assert g.stddev.unit is u.m
+
+
+def test_quantity_parameter_descriptors():
+    """
+    Test Model classes that specify units in their Parameter descriptors,
+    either via a Quantity default, or explicit use of the unit argument.
+    """
+
+    def tests(TestModel):
+        assert TestModel.a.unit == u.m
+        assert TestModel.a.default == 1.0
+
+        m = TestModel()
+        assert m.a.unit == u.m
+        assert m.a.default == m.a.value == 1.0
+
+        m = TestModel(2.0 * u.m)
+        assert m.a.unit == u.m
+        assert m.a.value == 2.0
+        assert m.a.default == 1.0
+
+        # Instantiate with a different, but compatible unit
+        m = TestModel(2.0 * u.pc)
+        assert m.a.unit == u.pc
+        assert m.a.value == 2.0
+        # The default is still in the original units
+        assert m.a.default == 1.0
```

Comment:
Or that its value should change when the unit changes. Though that seems a bit more fragile.